### PR TITLE
accept a list of {templateIds, query} as input to searchForever

### LINKED ISSUE #4073: union of queries
@hurryabit would like to have multiple queries running over a single websocket `searchForever` stream. Since we already handle multiple template IDs this way (by compiling the query for each template ID and then returning the union), providing more queries up-front would be relatively straightforward.

This union doesn't belong in the query language _itself_, because for a pair of {templateId..., q1 | q2}, q1 _and_ q2 must be well-typed for every templateId, but the idea here is to have multiple pairs {tid1..., q1}, {tid2..., q2}, for which q1 is only checked against tid1..., and likewise for 2.

These pairs will be specified up front ("static multiplex"); we may consider being able to send a message to an ongoing stream to add a new query to its results delivered from now on ("dynamic multiplex") separately from this issue. Static multiplexing fits well within our current framework, whereas dynamic multiplexing would be a significant rework.

In static multiplexing case:

* `search` endpoint _may_ also support the query format, so that `search` and `searchForever` continue to have exactly the same input format
* [x] should "which query passed" be identified, or is the template ID sufficient to disambiguate in practical circumstances? (yes, all queries that passed)

In dynamic multiplexing case:

* [ ] respond starting from the beginning or just from this point on?
* [ ] possible to cancel ongoing queries, leaving the other ones running?

### PR DESCRIPTION
**Pull request by S11001001**
The resulting search is the inclusive union of all pairs given. Fixes #4073.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: docs/source/json-api/index.rst [hunk 1]
```diff
@@ -801,10 +810,14 @@ to :doc:`lf-value-specification`::
             },
             "signatories": ["Alice"],
             "contractId": "#1:0",
-            "templateId": "6cc82609ede6576e5092e8c89a2c7a658efe15ae1347fc38eb316f787fa132bc:Iou:Iou"
-        }
+            "templateId": "b70bbfbc77a4790f66d4840cb19f657dd20848f5e2f64e39ad404a6cbd98cf75:Iou:Iou"
+        },
+        "matchedQueries": [1, 2]
```
Comment (S11001001) [inline docs/source/json-api/index.rst]: @hurryabit Currently `"matchedQueries"` are always emitted as demonstrated in docs even if only one query is supplied; please let me know if this is undesirable.

Comment (hurryabit) [reply docs/source/json-api/index.rst]: I think this is perfectly fine. Ignoring a field is no hard work.

File: ledger-service/http-json/src/main/scala/com/digitalasset/http/util/Collections.scala [hunk 2]
```diff
@@ -25,5 +25,23 @@ object Collections {
       }
       (esb.result, thatb.result)
     }
+
+    /*
+    def unzipMap[B, C, Bs, Cs](f: A => (B, C))(implicit bs: CanBuildFrom[Self, B, Bs], cs: CanBuildFrom[Self, C, Cs]) = {
```
Comment (leo-da) [inline ledger-service/http-json/src/main/scala/com/digitalasset/http/util/Collections.scala]: do you want to keep it?

File: ledger-service/http-json/src/main/scala/com/digitalasset/http/WebSocketService.scala [hunk 3]
```diff
@@ -50,66 +54,88 @@ object WebSocketService {
...
       JsArray(
-        step.deletes.iterator.map(inj("archived")).toVector
-          ++ Liskov.co[StepAndErrors, LfV, JsValue](lfv)(this).step.inserts.map(inj("created"))
-          ++ errors.map(inj[String]("error") compose (_.message)))
+        step.deletes.iterator.map(inj("archived", _)).toVector
+          ++ thisw.step.inserts.map {
+            case (ac, pos) =>
+              val acj = inj("created", ac)
+              acj copy (fields = acj.fields ++ pos)
+          } ++ errors.map(e => inj("error", e.message)))
```
Comment (leo-da) [inline ledger-service/http-json/src/main/scala/com/digitalasset/http/WebSocketService.scala]: why don't we want to emit `errors` first? the client may decide to cancel the stream processing entirely based on what error(s) they receive.

Comment (leo-da) [reply ledger-service/http-json/src/main/scala/com/digitalasset/http/WebSocketService.scala]: @hurryabit ^^^

Comment (S11001001) [reply ledger-service/http-json/src/main/scala/com/digitalasset/http/WebSocketService.scala:71]: This block assembles a single message; the client will get all this at once.

**Pull request merged**